src/esp/gfx/PbrShader.cpp

```diff
@@ -0,0 +1,484 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "PbrShader.h"
+
+#include <Corrade/Containers/Reference.h>
+#include <Corrade/Utility/Assert.h>
+#include <Corrade/Utility/Debug.h>
+#include <Corrade/Utility/DebugStl.h>
+#include <Corrade/Utility/FormatStl.h>
+#include <Corrade/Utility/Resource.h>
+#include <Magnum/GL/Context.h>
+#include <Magnum/GL/Extensions.h>
+#include <Magnum/GL/Shader.h>
+#include <Magnum/GL/Texture.h>
+#include <Magnum/GL/Version.h>
+#include <Magnum/ImageView.h>
+#include <Magnum/Math/Color.h>
+#include <Magnum/Math/Matrix3.h>
+#include <Magnum/Math/Matrix4.h>
+#include <Magnum/PixelFormat.h>
+
+#include "esp/core/esp.h"
+#include "esp/io/io.h"
+
+#include <sstream>
+
+// This is to import the "resources" at runtime. When the resource is
+// compiled into static library, it must be explicitly initialized via this
+// macro, and should be called *outside* of any namespace.
+static void importShaderResources() {
+  CORRADE_RESOURCE_INITIALIZE(ShaderResources)
+}
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace gfx {
+
+namespace {
+enum TextureUnit : uint8_t {
+  BaseColor = 0,
+  Roughness = 1,
+  Metallic = 2,
+  Normal = 3,
+  // NoneRoughnessMetallic or OcclusionRoughnessMetallic texture
+  Combined = 4,
+};
+}  // namespace
+
+PbrShader::Flags PbrShader::generateCorrectFlags(Flags originalFlags) {
+  Flags result = originalFlags;
+  // NOTE:
+  // The priority of different kind of textures is as follows (priority
+  // means if two textures with different priorities exist at the same time,
+  // shader will adopt the texture with the higher priority, and ignore the
+  // other one.)
+  // 0 (highest): OcclusionRoughnessMetallicTexture
+  // 1          : NoneRoughnessMetallicTexture
+  // 2          : RoughnessTexture, MetallicTexture
+  if (result & Flag::OcclusionRoughnessMetallicTexture) {
+    result &= ~Flag::NoneRoughnessMetallicTexture;
+    result &= ~Flag::RoughnessTexture;
+    result &= ~Flag::MetallicTexture;
+  } else if (result & Flag::NoneRoughnessMetallicTexture) {
+    result &= ~Flag::RoughnessTexture;
+    result &= ~Flag::MetallicTexture;
+  }
+
+  return result;
+}
```

Comment:
To avoid this kind of explicit postprocessing, what about having a flag like this:

```cpp
    // Explicitly saying "packed" to suggest that it already includes both
    PackedNoneRoughnessMetallicTexture = (1 << 3)|RoughnessTexture|MetallicTexture,
```

(and later for Occlusion as well, but you don't handle that one yet). Then any combination of enum values would be implicitly valid, and you'd only have to take extra care when passing the enum values through to the shader:

```cpp
// equiv. to (flags_ & Flag::PackedNoneRoughnessMetallicTexture) == Flag::PackedNoneRoughnessMetallicTexture
if(flags_ >= Flag::PackedNoneRoughnessMetallicTexture) {
  frag.addSource("#define NON_ROUGHNESS_METALLIC_TEXTURE\n");
} else {
  frag
      .addSource(flags_ & Flag::BaseColorTexture ? "#define BASECOLOR_TEXTURE\n"
                                                 : "")
      .addSource(flags_ & Flag::RoughnessTexture ? "#define ROUGHNESS_TEXTURE\n"
                                                 : "")
      .addSource(flags_ & Flag::MetallicTexture ? "#define METALLIC_TEXTURE\n"
                                                : "");
}
```

Same change for the `setUniform(uniformLocation("metallicTexture"), ...` block after.

Reply:
I think what I have done here is a cleaner way to handle this situation as it solves the "conflicts" once for all and it is cheap. I do not need to have any nested if clause any more anywhere, either in the `PbrShader`, or `PbrDrawable`. 
Later if you would like to change the logic, relations among the textures, or add new type of textures (such as independent occlusion), again, easy, you just need to change one place -- here, in this function.
Second, I am not a big fan of the operator like `>=`, as a user to your library, I am rather confused about this operator, what does it mean? The "flags" is a bitmask. I only need to query a bit (`&`), set a bit (`|=`) or reset a bit (`&=~`). That is it. `>=` looks really strange and not straightforward to me, honestly. Each time, I will have to think for a while, and check the "Flag" definition to get the answer.

Reply:
Depends :) I'd say having the API designed in a way that makes disallowed combinations impossible (and then having that all done and checked in a single central place) is better than having to explicitly scatter "fixup" calls everywhere *including* the call sites (like the `PbrDrawable`). Actually, seeing that fixup call in the drawable prompted me to write that comment -- because *what if* a similar check will need to be done in some other place, how many of such places will get forgotten?

About the `>=`, of course. That's precisely why I put the explanatory comment there -- I don't *force* you to use that, it's just a shorthand for the rather verbose check. It's the ⊇ operation from the set theory, and if you squint a bit, it's not *that* different from ≥. (Another equivalent, which is shorter, but adds even more congitive overhead is `!(a & ~b)`.)

Reply:
Lets say:

        enum class Flag: unsigned int {
            A = 1 << 0,
            B = 1 << 1,
            C = 1 << 2 | A,
            D = 1 << 3,
        };
`Flag foo = Flag|D`; So `foo` is 1000 which means `foo >= Flag|C` because Flag|C is 0101, but it does not necessarily include feature A and feature C. Correct? That is how I understand the operator `>=`.


 >  // Explicitly saying "packed" to suggest that it already includes both
 >   PackedNoneRoughnessMetallicTexture = (1 << 3)|RoughnessTexture|MetallicTexture,

You fully misunderstood what I meant here. Each bit in the bitmask is exclusive. This is different from your Magnum library. "RoughnessTexture" does NOT mean it has Roughness in the texture, but I do not care if it is in its own texture or a packed texture. It means roughness is stored in an independent, unique texture. And once my shader knows this fact, it will extract the roughness from R channel, not from G channel.

> because what if a similar check will need to be done in some other place, how many of such places will get forgotten?

Your method suffers the same problem. The same logic applies as well. Because you need the branching logic everywhere then, namely, you will have to repeat the nested "if clause" in your previous comments, for example, here, and in bindRoughnessTexture (because you will have to check if NONE_ROUGHNESS_METALLIC_TEXTURE or OCCLUSION_ROUGHNESS_METALLIC_TEXTURE are presented or not), bindMetallicTexture, and more places.

What I can improve is that, since shader key is, eventually, used in the shaderManager (besides `PbrShader` class). So I guess I will call such "fixup" in the getShaderKey function (and the PbrShader constructor of course.) 

> checked in a single central place

Depends on how you view this problem. on the contrary, I believe that is exactly what I did here. I checked and fixed the logic in a single central place -- generateCorrectFlags. And this a core idea I tried to convey in my first reply. After calling this function, it is worry-free, "if-clause" free.

> !(a & ~b)
This is a big cognitive overhead, again, not straightforward, not even close. So you are saying `a >= b`, it means b belongs to a. Is it correct? I did not see why it helps your claim (making the "disallowed combinations impossible") because you need the "if clause" to assist you everywhere. However, my "fixup" function did what exactly you claimed -- making the "disallowed combinations impossible".

Also, the burden for the developer is a bit heavier, she must understand that for such "combination bit", she should always remember to use ">=" instead of "&".

The fundamental difference on this problem between us is whether any bitmask should carry more than 1 meaning. I do not think so but you do.

Reply:
Okay, fair enough :) I'm not 100% sure yet how I will handle packed textures in builtin shaders, so this was just one of the ideas I had. Another is supplying an explicit swizzle for each texture, but that has its own bag of problems.

> That is how I understand the operator >=

In case of bit sets I understand it as "is a superset of" (that's why I said ⊇), while you understand it as "is numerically larger than" like with integers.

Reply:
> "is numerically larger than" like with integers.

I believe most developers will take it for granted. Especially, the type of the Flag element is set as `UnsignedInt`.  It is a strong hint. They will assume the "numerically larger than", because it is the most common way to understand it. So with all due respect, I am not convinced `>=` is a good design in the Magnum since it goes against the common sense, in my shallow opinion.